text/0000-combined-restorer-analyzer-phases.md

```diff
@@ -0,0 +1,130 @@
+# Meta
+[meta]: #meta
+- Name: Combined restorer/analyzer and cacher/exporter phases.
+- Start Date: 2019-08-02
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+Though not formally part of the spec, the project description and the
+lifecycle reference implementation contain the following four phases:
+
+- `/lifecycle/restorer` (considered an optional platform step)
+- `/lifecycle/analyzer`
+- `/lifecycle/exporter`
+- `/lifecycle/cacher` (considered an optional platform step)
+
+This proposal seeks to combine `restorer` with `analyzer` (to be named
+`restorer`) and `cacher` with `exporter` (to be named `exporter`).
+
+# Motivation
+[motivation]: #motivation
+
+In the current lifecycle reference implementation, the `restorer` phase
+fetches cached layers from a previous cache image. These layers can be
+arbitrarily large and, when using a remote repository, can represent a large
+amount of overhead. The `analyzer` phase runs after `restorer` and it
+may deterine that particular cache layers are no longer valid and remove them
+from the rest of the process - however, we've already paid a large cost to
+fetch them in the first place.
+
+The determination of cache layer validity can frequently be performed
+through the analysis of metadata: comparing metadata from the cache image with
+metadata from the previous application image. However, this comparison can
+only be performed with a combined `restorer` and `analyzer` phase.
+
+A parallel optimization potential exists when combining the `cacher` and
+`exporter` phases.
+
+# What it is
+[what-it-is]: #what-it-is
+
+## Combined `restorer` phase.
+
+The spec will formalize the (combined) `restorer` phase like this:
+
+```
+/lifecycle/restorer \
+  -group=[(required) group.toml file] \
+  -layers=[(required) layers directory] \
+  -image=[(optional) repository url for existing application image] \
+  -cache=[(optional) local directory, or repository url for cache image]
+```
+
+The `restorer` phase will be responsible for the combined activities of the
+previous `restorer` and `analyzer` phases
+(see https://github.com/buildpack/spec/blob/master/buildpack.md#phase-2-analysis).
+
+The previous `/lifecycle/analyzer` phase will be removed from the spec.
+
+
+## Combined `exporter` phase.
+
+The spec will formalize the (combined) `exporter` phase like this:
+
+```
+/lifecycle/exporter \
+  -app=[(required) the application source directory] \
+  -base=[(required) repository url for the base image to build on top of] \
+  -group=[(required) group.toml file] \
+  -layers=[(required) layers directory] \
+  -image=[(required) repository url for new application image] \
```

Comment:
We should perhaps call out in the spec that this value must be a tag reference when it is passed to exporter, where the analogous field in the `restorer` can refer to a specific concrete image (either a digest reference or an image ID in the daemon case)

Reply:
Also, the current reference lifecycle supports exporting to multiple tags provided they are on the same registry. We might want to spec that behavior.

Reply:
Opinions on (1) multiple `-image` tags, (2) delimited (e.g., comma-separated) `-image` tag, or (3) a series of positional args?

Reply:
I vote for option 3, but I'm open to option 1 if you want to make a case for it.

In my mind `<image>` makes sense as the only positional argument because it feels different than the other parameters. The other flags provide optional ways to configure `export`. Image is "the thing you are exporting" so it takes a different/primary position in the usage.

But I suspect someone could make an equally valid argument for the opposite position.